ihmc-robotics-toolkit/src/main/java/us/ihmc/robotics/math/filters/TransferFunctionDiscretizer.java

```diff
@@ -0,0 +1,204 @@
+
+   // Used for temporary private calculations.
+   private DMatrixRMaj outputVec;
+   private DMatrixRMaj tempVec;
+   private int numCols;
+
+   // Desired Transfer function numerator and denominator.
+   private DMatrixRMaj numerator;
+   private DMatrixRMaj denominator;
+
+   // Input/Output Coefficients.
+   private DMatrixRMaj inputCoefficients;
+   private DMatrixRMaj outputCoefficients;
+
+   private double k;
+   private double p;
+   private double fs;
+   private double output;
+   private int n;
+   private int m;
+   private String name;
+
+   public TransferFunctionDiscretizer(ContinuousTransferFunction tf, double sampleFrequency)
+   {
+      this(tf.getName(), tf, sampleFrequency);
+   }
+
+   /*
+    * @param name Name of the filter.
+    * @param k Transfer Function Gain
+    * @param numerator Order of continuous polynomial coefficients from largest to smallest order, i.e. [b_0, b_1, ... , b_{m-1}, b_{m}].
+    * @param demoninator Order of continuous polynomial coefficients from largest to smallest order, i.e. [a_0, a_1, ..., a_{n-1}, a_n].
+    * @param sampleFrequency Frequency [Hz] which this variable will be updated. Filter performance will be worse if this is not accurate.
+    */
+   public TransferFunctionDiscretizer(String name, ContinuousTransferFunction tf, double sampleFrequency)
+   {
+
+      this.name = name;
+
+      m = tf.getNumerator().length;
+      n = tf.getDenominator().length;
+
+      // Handles multiple cases of m and n for matrix sizes.
+      if (m > n)
+      {
+         throw new IllegalArgumentException("Transfer Function is non-causal!");
+      }
+      else if (m < n)
+      {
+         double[] d = new double[n];
+         System.arraycopy(tf.getNumerator(), 0, d, n - m, m);
+         this.numerator = new DMatrixRMaj(1, d.length, false, d);
+      }
+      else
+      {
+         this.numerator = new DMatrixRMaj(1, m, false, tf.getNumerator());
+      }
+
+      this.denominator = new DMatrixRMaj(1, n, false, tf.getDenominator());
+
+      this.k = tf.getGain();
+      this.fs = sampleFrequency;
+      this.p = 1 / (2 * this.fs);
+
+      // Apply gain across numerator.
+      CommonOps_DDRM.scale(k, this.numerator);
+
+      // Solve for the corresponding input/output filter coefficients.
+      buildFilter();
+   }
+
+   private void buildFilter()
+   {
+
+      numerator = solveFx_to_Fz(numerator);
+      denominator = solveFx_to_Fz(denominator);
+
+      // Solve for outputCoefficients.
+      inputCoefficients = numerator.copy();
+      CommonOps_DDRM.scale(1.0 / denominator.get(0), inputCoefficients);
+
+      // Solve for inputCoefficients.
+      double[] d = new double[n - 1];
+      System.arraycopy(denominator.data, 1, d, 0, denominator.numCols - 1);
+      outputCoefficients = new DMatrixRMaj(1, d.length, false, d);
+      CommonOps_DDRM.scale(-1.0 / denominator.get(0), outputCoefficients);
+   }
+
+   private double calc;
+
+   private DMatrixRMaj solveFx_to_Fz(DMatrixRMaj Fx)

```

Comment:
this is full of garbage creation. Please fix it. I also have no idea what this method does.